### PR TITLE
Feat/multiple req keys

### PR DESCRIPTION
**Pull request by matiu**
support multiple request pub keys per copayer (see https://github.com/bitpay/copay/issues/3044)

### CODE REVIEW

File: lib/errors/errordefinitions.js [hunk 1]
```diff
@@ -15,8 +15,10 @@ var errors = {
   INSUFFICIENT_FUNDS: 'Insufficient funds',
   INSUFFICIENT_FUNDS_FOR_FEE: 'Insufficient funds for fee',
   INVALID_ADDRESS: 'Invalid address',
+  KEY_IN_COPAYER: 'Key already registered',
   LOCKED_FUNDS: 'Funds are locked by pending transaction proposals',
   NOT_AUTHORIZED: 'Not authorized',
+  TO_MANY_KEYS: 'To many keys registered',
```
Comment (isocolsky) [inline lib/errors/errordefinitions.js]: Small typo, should be `TOO_MANY_KEYS`

File: lib/server.js [hunk 2]
```diff
@@ -373,6 +323,110 @@ WalletService.prototype._notify = function(type, data, opts, cb) {
...
+      copayerId: copayer.id,
+      wallet: wallet
+    });
+  });
+};
+
+/**
+ * Adds access to a given copayer
+ *
+ * @param {Object} opts
```
Comment (isocolsky) [inline lib/server.js]: Would it make sense to accept a text field as an identifier for this particular access key? This way we could add access keys for agents.

Comment (matiu) [inline lib/server.js]: Definitely. Will add it.

**Comment by matiu** [comment]
updated, ready for review.

Needs: https://github.com/bitpay/bitcore-wallet-utils/pull/30

**Comment by matiu** [comment]
updated with @isocolsky comments. BWC related commit (https://github.com/bitpay/bitcore-wallet-client/pull/107) updated with @Kirvx comment.

thanks!

File: lib/model/copayer.js [hunk 3]
```diff
@@ -30,16 +31,42 @@ Copayer.create = function(opts) {
 
   x.id = WalletUtils.xPubToCopayerId(x.xPubKey);
   x.name = opts.name;
-  x.signature = opts.signature; // So third parties can check independently
+
+  if (!x.requestPubKeys) {
```
Comment (isocolsky) [inline lib/model/copayer.js]: Did you mean `opts.requestPubKeys` instead of `x.requestPubKeys`?

Comment (matiu) [inline lib/model/copayer.js]: fixed, thanks.

File: lib/model/wallet.js [hunk 4]
```diff
@@ -94,7 +94,8 @@ Wallet.prototype.isShared = function() {
 
 Wallet.prototype._updatePublicKeyRing = function() {
   this.publicKeyRing = _.map(this.copayers, function(copayer) {
-    return _.pick(copayer, ['xPubKey', 'requestPubKey', 'isTemporaryRequestKey']);
+    return _.pick(copayer, ['xPubKey', 'requestPubKey']);
+//    return _.pick(copayer, ['xPubKey']);
```
Comment (isocolsky) [inline lib/model/wallet.js]: :deciduous_tree: ...

File: lib/model/wallet.js [hunk 5]
```diff
@@ -107,20 +108,19 @@ Wallet.prototype.addCopayer = function(copayer) {
...
-  $.checkState(c)
-    .checkState(c.isTemporaryRequestKey);
+  var c = this.getCopayer(copayerId);
 
-  c.requestPubKey = requestPubKey;
-  c.isTemporaryRequestKey = false;
-  c.signature = signature;
-  this._updatePublicKeyRing();
+  //new ones go first
+  c.requestPubKeys.unshift({
```
Comment (isocolsky) [inline lib/model/wallet.js]: So, what about the first key (the one used to create the copayer). Will it be different in that it won't have additional attributes?

**Comment by matiu** [comment]
updated.

File: lib/server.js (hunk 2 above)
Comment (matiu) [inline lib/server.js]: done

**Comment by isocolsky** [comment]
Great job!!

**Comment by troggy** [comment]
@matiu My BWS instance started throwing 'Invalid proposal signature' because ``requestPubKeys`` is null [here](https://github.com/bitpay/bitcore-wallet-service/pull/311/files#diff-c945a46d13b34fcaff544d966cffcabaR180) for existing copayers. To fix I had to drop BWS database and clients had to reregister their wallets. Is it the only way or I miss something?

**Comment by isocolsky** [comment]
@troggy thank you for reporting. We could reproduce the issue. It was due to a bug reading the version number of the copayer object stored in the db. It should be fixed by #338

**Pull request merged**